[EFM] Add Epoch Fallback Phase

jordanschalm opened the pull request:
This PR adds a new epoch phase to the data model, representing a state where epoch fallback mode has been triggered **and** we are within the view range of the latest committed epoch. 

This phase is subtly different than the existing `EpochFallbackTriggered` flag in the protocol state:
- `EpochFallbackTriggered` may be true while in `EpochPhaseCommitted` (in this case the protocol state flag and the epoch phase are "unsynchronized")
- If current phase is not `EpochPhaseCommitted`, then when `EpochFallbackTriggered` is true, we must be in `EpochPhaseFallback`

AlexHentschel commented on state/protocol/badger/snapshot.go:
```diff
@@ -413,7 +413,7 @@ func (q *EpochQuery) Next() protocol.Epoch {
 	entry := epochState.Entry()
 
 	// if we are in the staking phase, the next epoch is not setup yet
```
```suggestion
	// if we are in the staking phase or fallback phase, the next epoch is not setup yet
```

AlexHentschel reviewed (approved):
Nice work. Looks good.

AlexHentschel commented:
Overall, I have one request to properly document the convention when we have the `EpochPhaseFallback`. My understanding is that we are in `EpochPhaseFallback` if and only if we would accept an Recovery event (the bocks where we enter EFM and exit EFM might be an exception from this rule, not sure).  I think we should document this in detail in various places in the code base:
* In particular, I think it is important to clearly document our convention when `EpochPhaseFallback` in the logic that controls the Epoch state during EFM https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/protocol_state/epochs/fallback_statemachine.go#L31-L40 together with the logic that _interprets_ the Epoch State and deduces the `EpochPhase`: https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/model/flow/protocol_state.go#L327-L329
* Furthermore, I would suggest to just copy this documentation to 
   * the interface: https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/protocol_state.go#L52-L53 
   * as well as in the adapter: https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/inmem/epoch_protocol_state.go#L99-L102
   * for consistency, I would be inclined to rename `Phase()` to `EpochPhase` in the [`Snapshot` interface](https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/snapshot.go#L131-L133) and `Snapshot` implementation [here](https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/inmem/snapshot.go#L81) and [here](https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/badger/snapshot.go#L84)


For now, I would just drop the following documentation into all the different places mentioned above. I have tried to not only document the epoch phases but also when the notifications are emitted. Please check that it is correct. While it is a lot of words, I think it is super important to unambiguously specify our convention (both for people using the code as well as for engineers attempting to extend the code later to not break anything). 
```golang
// EpochPhase returns the current epoch phase. During normal operations, each Epoch transitions through
// the phases:
//
//	║                                       Epoch N
//	║       ╭─────────────────────────────────┴─────────────────────────────────╮
//	║   finalize view            EpochSetup            EpochCommit
//	║     in epoch N            service event         service event
//	║        ⇣                       ⇣                     ⇣
//	║        ┌─────────────────┐     ┌───────────────┐     ┌───────────────────┐
//	║        │EpochPhaseStaking├─────►EpochPhaseSetup├─────►EpochPhaseCommitted├ ┄>
//	║        └─────────────────┘     └───────────────┘     └───────────────────┘
//	║        ⇣                       ⇣                     ⇣
//	║   EpochTransition     EpochSetupPhaseStarted    EpochCommittedPhaseStarted
//	║    Notification            Notification               Notification
//
// However, if the Protocol State encounters any unexpected epoch service events, or the subsequent epoch
// fails to be committed by the `EpochCommitSafetyThreshold`, then we enter Epoch Fallback Mode [EFM].
// Depending on whether the subsequent epoch has already been committed, the EFM progress differs slightly.
// In a nutshell, we always enter the _latest_ epoch already committed on the happy path (if there is any)
// and the follow the fallback protocol.
//
// Scenario A: the future Epoch N is already committed, when we enter EFM
//
//	║      Epoch N-1                            Epoch N
//	║   ···──┴─────────────────────────╮ ╭─────────────┴───────────────────────────────────────────────╮
//	║      invalid service                finalize view                   EpochRecover
//	║            event                    in epoch N                      service event
//	║              ⇣                      ⇣                    ┊          ⇣
//	║     ┌──────────────────────────┐    ┌────────────────────┊────┐     ┌───────────────────────────┐
//	║     │   EpochPhaseCommitted    ├────►    EpochPhaseFallback   ├─────►    EpochPhaseCommitted    ├ ┄>
//	║     └──────────────────────────┘    └────────────────────┊────┘     └───────────────────────────┘
//	║              ⇣                      ⇣                    ┊          ⇣
//	║   EpochFallbackModeTriggered     EpochTransition   EpochExtended*   EpochFallbackModeExited
//	║          Notification             Notification      Notification    + EpochCommittedPhaseStarted Notifications
//	║              ┆                                                      ┆
//	║              ╰┄┄┄┄┄┄┄┄┄┄ EpochFallbackTriggered is true ┄┄┄┄┄┄┄┄┄┄┄┄╯
//
// With 'EpochExtended*' we denote that there can be zero, one, or more Epoch Extension (depending on when
// we receive a valid EpochRecover service event.
//
// Scenario B: we are in Epoch N without any subsequent epoch being committed when entering EFM
//
//	║                         Epoch N
//	║ ···────────────────────────┴───────────────────────────────────────────────────────────────╮
//	║              invalid service event or                         EpochRecover
//	║         EpochCommitSafetyThreshold reached                    service event
//	║                           ⇣                      ┊            ⇣
//	║  ┌────────────────────┐   ┌──────────────────────┊──────┐     ┌───────────────────────────┐
//	║  │ EpochPhaseStaking  │   │     EpochPhaseFallback      │     │   EpochPhaseCommitted     │
//	║  │ or EpochPhaseSetup ├───►                      ┊      ├─────►                           ├ ┄>
//	║  └────────────────────┘   └──────────────────────┊──────┘     └───────────────────────────┘
//	║                           ⇣                      ┊            ⇣
//	║            EpochFallbackModeTriggered     EpochExtended*      EpochFallbackModeExited
//	║                     Notification           Notification       + EpochCommittedPhaseStarted Notifications
//	║                           ┆                                   ┆
//	║                           ╰┄┄ EpochFallbackTriggered true ┄┄┄┄╯
func (e *EpochProtocolStateEntry) EpochPhase() EpochPhase {
```

AlexHentschel commented on model/flow/protocol_state.go:
```diff
@@ -327,6 +327,16 @@ func (e *RichEpochProtocolStateEntry) CurrentEpochFinalView() uint64 {
 // EpochPhase returns the current epoch phase.
 // The receiver EpochProtocolStateEntry must be properly constructed.
 func (e *EpochProtocolStateEntry) EpochPhase() EpochPhase {
+	if e.EpochFallbackTriggered {
```
I think it would be worth-while to add a comment here to call out the dependency on the Epoch Fallback state machine. just to reduce the risk of code changes in the future accidentally breaking the consistency between both components. 

```suggestion
	// CAUTION: the logic below that deduces the EpochPhase must be consistent with `epochs.FallbackStateMachine`,
	// which sets the fields we are using here. Specifically, we require that the FallbackStateMachine clears out 
	// any tentative values for a subsequent epoch _unless_ that epoch is already committed. 
	if e.EpochFallbackTriggered {
```

jordanschalm commented:
> Furthermore, I would suggest to just copy this documentation to [...] the interface https://github.com/onflow/flow-go/blob/f88fe607cde248eee6ca6263795c2b0e098e15b1/state/protocol/protocol_state.go#L52-L53

I added this to `EpochProtocolStateAdapter`, but not the interface. Saying the concrete implementation must be properly constructed in an abstract interface definition didn't make much sense to me; correct construction of a specific struct type is the domain of the implementation, not the interface. 

> For now, I would just drop the following documentation into all the different places mentioned above.

Thank you for the excellent in-depth documentation. I have incorporated this into the `flow.EpochPhase` godoc, along with the existing phase change diagram. Elsewhere, I added links to the `flow.EpochPhase` type and corresponding godoc rather than copying the ~50 lines directly.
